Realm/Realm/DatabaseTypes/RealmCollectionBase.cs

```diff
@@ -189,15 +190,41 @@ public IRealmCollection<T> Freeze()
             return CreateCollection(frozenRealm, frozenHandle);
         }
 
-        public IDisposable SubscribeForNotifications(NotificationCallbackDelegate<T> callback)
-            => SubscribeForNotificationsImpl(callback, shallow: false);
+        public IDisposable SubscribeForNotifications(NotificationCallbackDelegate<T> callback, KeyPathsCollection? keyPathsCollection = null)
+        {
+            keyPathsCollection ??= KeyPathsCollection.Full;
+
+            if (keyPathsCollection.Type == KeyPathsCollectionType.Explicit && !ContainsRealmObjects())
+            {
+                throw new InvalidOperationException("Key paths can be used only with collections of Realm objects");
+            }
+
+            return SubscribeForNotificationsImpl(callback, keyPathsCollection);
+        }
 
-        internal IDisposable SubscribeForNotificationsImpl(NotificationCallbackDelegate<T> callback, bool shallow)
+        internal IDisposable SubscribeForNotificationsImpl(NotificationCallbackDelegate<T> callback, KeyPathsCollection keyPathsCollection)
         {
             Argument.NotNull(callback, nameof(callback));
-            _notificationCallbacks.Value.Add(callback, shallow);
 
-            return NotificationToken.Create(callback, c => UnsubscribeFromNotifications(c, shallow));
+            if (keyPathsCollection.Type == KeyPathsCollectionType.Explicit)
+            {
+                var managedResultsHandle = GCHandle.Alloc(this, GCHandleType.Weak);
+                var callbackHandle = GCHandle.Alloc(callback, GCHandleType.Weak);
+
+                var token = Handle.Value.AddNotificationCallback(GCHandle.ToIntPtr(managedResultsHandle), keyPathsCollection,
+                    GCHandle.ToIntPtr(callbackHandle));
+
+                return NotificationToken.Create(callback, _ => token.Dispose());
+            }
+
+            // For notifications with type Default or Shallow we cache the callbacks on the managed level, to avoid creating multiple notifications in core
```

Comment:
```suggestion
            // For notifications with type Full or Shallow we cache the callbacks on the managed level, to avoid creating multiple notifications in core
```